ExtendedProcessFailedEventArgs.md (#3824)
MicrosoftEdge/WebView2Feedback: specs/ExtendedProcessFailedEventArgs.md

> Title

Reviewer: Please update title

===

# Background

> Code integrity is a feature of Windows that verifies the authenticity and integrity of the code that runs on the system. It helps protect it from malware, tampering, and unauthorized changes. Code integrity checks the digital signatures of the files that are loaded into memory, and prevents any file that does not have a valid signature from running in WebView2. We are extending ProcessFailedEventArgs with BlockedFile property which caused webview2 process to exit with code STATUS_INVALID_IMAGE_HASH.

Reviewer: Please ensure lines aren't longer than 80 or 100 characters. Its hard for reviewers reading on github otherwise

# Examples

    ```c#
        /// This is an event handler for our CoreWebView2's ProcessFailedEvent
        private void CoreWebView2_ProcessFailed(object sender, CoreWebView2ProcessFailedEventArgs e)
        {
            if (e.ExitCode == -1073740760 /*STATUS_INVALID_IMAGE_HASH*/)
            {

>                 SendTelemetry(e.BlockedFile);

Reviewer: Would be nice to add a comment here (and in the corresponding spot in the other sample code) that explains what its doing. eg:
```suggestion
                // If the renderer process crashed because of STATUS_INVALID_IMAGE_HASH, then we
                // want to log to our app's telemetry the name of the DLL that caused the issue.
                SendTelemetry(e.BlockedFile);
```

            }
        }
    ```

    ```cpp
    CHECK_FAILURE(m_webView->add_ProcessFailed(
        Callback<ICoreWebView2ProcessFailedEventHandler>(
            [this](ICoreWebView2* sender, ICoreWebView2ProcessFailedEventArgs* argsRaw)
                -> HRESULT {
                wil::com_ptr<ICoreWebView2ProcessFailedEventArgs> args = argsRaw;
                int exit_code;
                CHECK_FAILURE(args->get_ExitCode(&exit_code));

                if (exit_code == -1073740760 /*STATUS_INVALID_IMAGE_HASH*/) {
                    wil::unique_cotaskmem_string blockedFile;
                    CHECK_FAILURE(arg_blocked_file->get_BlockedFile(&blockedFile));

                    SendTelemetry(blockedFile);
                }

                return S_OK;
            }
    ```

# API Details

```
/// A continuation of the ICoreWebView2ProcessFailedEventArgs2 interface

> /// fot getting blocked file for code integrity process failures.

Reviewer: Typo in this line. But also you can just leave off this part of the description and leave as generic 'A continuation of ... interface' since we may end up having to merge these sorts of interfaces together later.

```suggestion
```

[uuid(a9fc1af8-f934-4f0f-a788-7be0808c329b), object, pointer_default(unique)]

> interface ICoreWebView2ProcessFailedEventArgs : IUnknown {

Reviewer: The first sentence of the ref docs gets pulled out as the brief description of the API for at least one of our ref doc systems. So its good to have a first sentence that sort of summarizes the API:

```suggestion
interface ICoreWebView2ProcessFailedEventArgs : IUnknown {
    /// This property is the full path of the module that caused the renderer process crash in cases of Windows Code Integrity failures.
```

>     /// Code Integrity is a feature that verifies the integrity and

Reviewer: Would be nice to make this a link to more info about this Windows feature. Do you have a better URI for documentation of Windows Code Integrity?

```suggestion
    /// [Windows Code Integrity](https://learn.microsoft.com/en-us/mem/intune/user-help/you-need-to-enable-code-integrity) is a feature that verifies the integrity and
```

    /// authenticity of dynamic-link libraries (DLLs)
    /// on Windows systems. It ensures that only trusted
    /// code can run on the system and prevents unauthorized or
    /// malicious modifications.
    /// When ProcessFailed occurred due to a failed Code Integrity check,

>     /// this property returns the name of the blocked file that was prevented from

Reviewer: Name or full path? Full path is nicer since it would allow the host app to inspect the file to produce a hash or examine the file's cert to better identify the DLL.

```suggestion
    /// this property returns the full path of the blocked file that was prevented from
```

    /// loading on the system.
    /// The webview2 process which tried to load blocked DLL will fail with
    /// exit code STATUS_INVALID_IMAGE_HASH(-1073740760).
    /// A file can be blocked for various
    /// reasons, such as:
    /// - It has an invalid or missing signature that does
    /// not match the publisher or signer of the file.
    /// - It has been tampered with or corrupted by malware or other software.
    /// - It has been blocklisted by an administrator or a security policy.

>     /// This property always will be empty if failure is not caused by

Reviewer: ```suggestion
    /// This property always will be the empty string if failure is not caused by
```

    /// STATUS_INVALID_IMAGE_HASH.

>     [propget] HRESULT BlockedFile([out, retval] LPWSTR* blockedFile);

Reviewer: Name thoughts:

Instead of `Blocked` we could blame it more like `FailureSource`, `FailureCause`
Or `Disallowed` or `Denied` I see in context of this feature rather than blocked but those are basically synonyms - not sure that's better.

Instead of `File` we could be more specific and say `Module` or `Dll` or `Image`. Of those Module is probably more precise than Dll and less ambiguous than Image.

And then this property isn't the file itself, its either the name or path of the file. So would be good to add that as a suffix.

How about:
`FailureSourceModulePath`

Code integrity failure source module path
===

# Background
[Windows Code Integrity](https://learn.microsoft.com/en-us/mem/intune/user-help/you-need-to-enable-code-integrity) is a feature that verifies the
integrity of the code that runs on the system. It helps protect it from malware,
tampering, and unauthorized changes. Code integrity checks the digital
signatures of the files that are loaded into memory, and prevents any
file that does not have a valid signature from running in WebView2 process.
We are extending ProcessFailedEventArgs with FailureSourceModulePath property
which caused webview2 process to exit with code STATUS_INVALID_IMAGE_HASH.

# Examples

>     ```c#

Reviewer 2: Remove the leading spaces here to let the backticks get interpreted correctly so this shows as a formatted code block.

Reviewer: Please fix.

        /// This is an event handler for our CoreWebView2's ProcessFailedEvent
        private void CoreWebView2_ProcessFailed(object sender, CoreWebView2ProcessFailedEventArgs e)
        {
            if (e.ExitCode == -1073740760 /*STATUS_INVALID_IMAGE_HASH*/)
            {
                // If the process crashed because of STATUS_INVALID_IMAGE_HASH,

>                 // then we want to log to our app's telemetry the name of the

Reviewer 3: Nit:
```suggestion
                // then we want to log to our app's telemetry the path of the
```

Reviewer: Please fix.

                // DLL that caused the issue.
                SendTelemetry(e.FailureSourceModulePath);
            }
        }
    ```

    ```cpp
    CHECK_FAILURE(m_webView->add_ProcessFailed(
        Callback<ICoreWebView2ProcessFailedEventHandler>(
            [this](ICoreWebView2* sender, ICoreWebView2ProcessFailedEventArgs* argsRaw)
                -> HRESULT {
                wil::com_ptr<ICoreWebView2ProcessFailedEventArgs> args = argsRaw;
                int exit_code;

>                 CHECK_FAILURE(args->get_ExitCode(&exit_code));

Reviewer 4: get_ExitCode is a method on ICoreWebView2ProcessFailedEventArgs2, so would have to QI to it first.

Reviewer: Please fix.

>                 if (exit_code == -1073740760 /*STATUS_INVALID_IMAGE_HASH*/) {

Reviewer 5: FailureSourceModulePath doesn't appear to be specific to STATUS_INVALID_IMAGE_HASH or Windows Code Integrity.  Is the expectation that this string could be populated in other failure cases, but STATUS_INVALID_IMAGE_HASH is the only case when we're currently promising it will always be populated?

Reviewer: Yes that is exactly what I was thinking.

                    wil::unique_cotaskmem_string modulePath;
                    CHECK_FAILURE(args->get_FailureSourceModulePath(&modulePath));

                    // If the process crashed because of STATUS_INVALID_IMAGE_HASH,
                    // then we want to log to our app's telemetry the name of the
                    // DLL that caused the issue.
                    SendTelemetry(modulePath);
                }

                return S_OK;
            }
    ```

# API Details

```
[uuid(a9fc1af8-f934-4f0f-a788-7be0808c329b), object, pointer_default(unique)]

> interface ICoreWebView2ProcessFailedEventArgs : IUnknown {

Reviewer 4: Um, [we already have a ICoreWebView2ProcessFailedEventArgs](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2processfailedeventargs?view=webview2-1.0.2045.28). and even a ICoreWebView2ProcessFailedEventArgs2. Shouldn't this be ICoreWebView2ProcessFailedEventArgs3?

Reviewer: Please fix. Yes, should be ICoreWebView2ProcessFailedEventArgs3 and it should be derived from ICoreWebView2ProcessFailedEventArgs2

    /// This property is the full path of the module that caused the
    /// crash in cases of Windows Code Integrity failures.
    /// [Windows Code Integrity](https://learn.microsoft.com/en-us/mem/intune/user-help/you-need-to-enable-code-integrity)
    /// is a feature that verifies the integrity and
    /// authenticity of dynamic-link libraries (DLLs)
    /// on Windows systems. It ensures that only trusted
    /// code can run on the system and prevents unauthorized or
    /// malicious modifications.
    /// When ProcessFailed occurred due to a failed Code Integrity check,
    /// this property returns the full path of the file that was prevented from
    /// loading on the system.
    /// The webview2 process which tried to load the DLL will fail with
    /// exit code STATUS_INVALID_IMAGE_HASH(-1073740760).
    /// A file can fail integrity check for various

>     /// reasons, such as:

Reviewer 3: May be out of scope, but will the developer ever need to know the more specific reason?

Reviewer: I don't believe this is required and we can look into adding it if necessary in the future. @vbryh-msft or @aluhrs13 can comment otherwise.

    /// - It has an invalid or missing signature that does
    /// not match the publisher or signer of the file.
    /// - It has been tampered with or corrupted by malware or other software.

>     /// - It has been blocklisted by an administrator or a security policy.

Reviewer 4: ```suggestion
    /// - It has been blocked by an administrator or a security policy.
```

Reviewer: Please fix.
